### PR TITLE
fix: improve the tactic for brecOn for inductive predicates.

### PR DESCRIPTION
**Pull request by DanielFabian**
This is an improved tactic for proving `brecOn` that correctly also works for `Acc`. It applies the slightly more carefully, which causes meta variables to be solved better. Furthermore, it applies backwards chaining because sometimes, e.g. in `Acc`, you need a multi-step argument.

### CODE REVIEW

File: src/Lean/Meta/IndPredBelow.lean [hunk 1]
```diff
@@ -265,11 +264,14 @@ where
...
-      let ctor := ctor.constName!.updatePrefix below.constName!
+      (←getMVarType m).withApp fun below args => 
       withMVarContext m do
-      let ctor := mkConst ctor below.constLevels!
+      args.back.withApp fun ctor ctorArgs => do
+      let ctorName := ctor.constName!.updatePrefix below.constName!
+      let ctor := mkConst ctorName below.constLevels!
+      let ctorInfo ← getConstInfoCtor ctorName
+      let (mvars, _, t) ← forallMetaTelescope ctorInfo.type
+      let ctor := mkAppN ctor $ mvars
```
Comment (leodemoura) [inline src/Lean/Meta/IndPredBelow.lean:274]: unnecessary `$`

File: src/Lean/Meta/IndPredBelow.lean [hunk 2]
```diff
@@ -279,31 +281,33 @@ where
   closeGoal (vars : BrecOnVariables) (m : MVarId) : MetaM Unit := do
     unless ←isExprMVarAssigned m do
       let m ← introNPRec m
-      unless ←solveByAssumption m do
-        let subGoals ← applyIH m vars
-        subGoals.forM (closeGoal vars)
+      unless ←backwardsChaining m 10 do
```
Comment (leodemoura) [inline src/Lean/Meta/IndPredBelow.lean:284]: Add option for configuring this limit.

File: src/Lean/Meta/IndPredBelow.lean [hunk 3]
```diff
@@ -279,31 +281,33 @@ where
...
+        withMVarContext m do
+        throwError "couldn't solve by backwards chaining: {MessageData.ofGoal m}"
```
Comment (leodemoura) [inline src/Lean/Meta/IndPredBelow.lean:286]: We could mention the depth threshold in the error message, and how the option name for changing it.

File: src/Lean/Meta/IndPredBelow.lean (hunk 1 above)
Comment (DanielFabian) [reply src/Lean/Meta/IndPredBelow.lean]: done

File: src/Lean/Meta/IndPredBelow.lean (hunk 3 above)
Comment (DanielFabian) [reply src/Lean/Meta/IndPredBelow.lean]: improved the message.

File: src/Lean/Meta/IndPredBelow.lean (hunk 2 above)
Comment (DanielFabian) [reply src/Lean/Meta/IndPredBelow.lean]: done

**Pull request merged**